File: app/src/main/java/org/test/koltintest/ShapeFragment.java
package org.test.koltintest;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

public class ShapeFragment extends Fragment {
    private final int layouts[] = {
            R.layout.fragment_first,
            R.layout.fragment_second,
            R.layout.fragment_third
    };

    private static final String ARG_INDEX = "index";

    public static ShapeFragment create() {
        return create(0);
    }

    public static ShapeFragment create(int index) {
        Bundle args = new Bundle();
        args.putInt(ARG_INDEX, index);

        ShapeFragment fragment = new ShapeFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        int index = getArguments().getInt(ARG_INDEX); // TODO código duplicado
        return inflater.inflate(layouts[index] /* */, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        view.setOnClickListener($ -> {
            int index = getArguments().getInt(ARG_INDEX); // TODO código duplicado
            int nextIndex = (index + 1) % layouts.length;
            String tag = "shape" + index;

            getFragmentManager().popBackStack(tag, FragmentManager.POP_BACK_STACK_INCLUSIVE);
            getFragmentManager().beginTransaction() // TODO podría hacerse genérico
                    .replace(getId(), ShapeFragment.create(nextIndex))
                    .addToBackStack(tag)
                    .commit();
        });
    }

}

File: app/src/main/java/org/test/koltintest/ShapeFragment.kt
package org.test.koltintest

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import org.test.replaceWith

class ShapeFragment : Fragment() {

    private val layouts = intArrayOf(
            R.layout.fragment_first,
            R.layout.fragment_second,
            R.layout.fragment_third)

    private val index by lazy { arguments!!.getInt(ARG_INDEX) }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) =
            inflater.inflate(layouts[index], container, false)!!

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        view.setOnClickListener {
            val nextIndex = (index + 1) % layouts.size
            val tag = "shape$index"

            fragmentManager!!.popBackStack(tag, FragmentManager.POP_BACK_STACK_INCLUSIVE)
            replaceWith(ShapeFragment.create(nextIndex), tag)
        }
    }

    companion object {
        private const val ARG_INDEX = "index"

        fun create(index: Int = 0) = ShapeFragment().apply {
            arguments = bundleOf(ARG_INDEX to index)
        }
    }

}
